services/api/src/api/routes/rows.py

```diff
@@ -0,0 +1,256 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2022 The HuggingFace Authors.
+
+import logging
+from functools import lru_cache, partial
+from typing import Any, List, Optional, Tuple
+
+import numpy as np
+import pyarrow as pa
+import pyarrow.parquet as pq
+from hffs.fs import HfFileSystem
+from libcommon.processing_graph import ProcessingStep
+from starlette.requests import Request
+from starlette.responses import Response
+from tqdm.contrib.concurrent import thread_map
+from typing_extensions import Protocol
+
+from api.authentication import auth_check
+from api.prometheus import StepProfiler
+from api.routes.endpoint import get_cache_entry_from_steps
+from api.utils import (
+    ApiCustomError,
+    Endpoint,
+    InvalidParameterError,
+    MissingRequiredParameterError,
+    UnexpectedError,
+    are_valid_parameters,
+    get_json_api_error_response,
+    get_json_ok_response,
+)
+
+MAX_ROWS = 100
+
+PARQUET_REVISION = "refs/convert/parquet"
+
+# TODO: manage private/gated datasets
+
+
+class FileSystemError(Exception):
+    pass
+
+
+class ParquetResponseFormatError(Exception):
+    pass
+
+
+class ParquetResponseEmptyError(Exception):
+    pass
+
+
+# TODO: how to invalidate the cache when the parquet branch is created or deleted?
+@lru_cache(maxsize=128)
+def get_parquet_fs(dataset: str, hf_token: Optional[str]) -> HfFileSystem:
+    """Get the parquet filesystem for a dataset.
+
+    The parquet files are stored in a separate branch of the dataset repository (see PARQUET_REVISION)
+
+    Args:
+        dataset (str): The dataset name.
+        hf_token (Optional[str]): The token to access the filesystem.
+
+    Returns:
+        HfFileSystem: The parquet filesystem.
+    """
+    return HfFileSystem(dataset, repo_type="dataset", revision=PARQUET_REVISION, token=hf_token)
+
+
+RowGroupReader = partial[Any]
+
+Index = Tuple[Any, List[RowGroupReader]]
+
+
+class Indexer(Protocol):
+    def __call__(self, dataset: str, config: str, split: str) -> Index:
+        ...
+
+
+def create_index(
+    config_parquet_processing_steps: List[ProcessingStep],
+    init_processing_steps: List[ProcessingStep],
+    hf_endpoint: str,
+    hf_token: Optional[str] = None,
+) -> Indexer:
+    @lru_cache(maxsize=128)
+    def index(
+        dataset: str,
+        config: str,
+        split: str,
+    ) -> Index:
+        with StepProfiler(method="rows.index", step="all"):
```

Comment:
Maybe to keep consistency with profiler in processing_step_endpoint method:
```suggestion
        with StepProfiler(method="index", step="all", context="endpoint: rows"):
```
Or something like that and same for other StepProfiler calls in this method